internal/service/resources/collector/collector.go

```diff
@@ -0,0 +1,229 @@
+			return nil
+		}
+	}
+}
+
+// init runs the onetime initialization steps for the collector
+func (c *Collector) init(ctx context.Context) error {
+	slog.Info("initializing data collector")
+
+	for _, d := range c.dataSources {
+		err := c.initDataSource(ctx, d)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// initDataSource initializes a single data source from persistent storage.  This recovers its unique UUID and
+// generation ID so that it continues from its last save point.
+func (c *Collector) initDataSource(ctx context.Context, dataSource DataSource) error {
+	name := dataSource.Name()
+	record, err := c.repository.GetDataSourceByName(ctx, name)
+	switch {
+	case errors.Is(err, utils.ErrNotFound):
+		// Doesn't exist so create it now.
+		result, err := c.repository.CreateDataSource(ctx, &models.DataSource{
+			Name:         name,
+			GenerationID: dataSource.GetGenerationID(),
+		})
+		if err != nil {
+			return fmt.Errorf("failed to create new data source %q: %w", name, err)
+		}
+
+		dataSource.SetID(*result.DataSourceID)
+		slog.Info("created new data source", "name", name, "uuid", *result.DataSourceID)
+	case err != nil:
+		return fmt.Errorf("failed to get data source %q: %w", name, err)
+	default:
+		dataSource.SetID(*record.DataSourceID)
+		dataSource.SetGenerationID(record.GenerationID)
+		slog.Info("restored data source",
+			"name", name, "uuid", record.DataSourceID, "generation", record.GenerationID)
+	}
+	return nil
+}
+
+// execute runs a single iteration of the main loop.  It does not return an error because all errors should be handled
+// gracefully.  If a truly unrecoverable error happens then a panic should be used to restart the process.
+func (c *Collector) execute(ctx context.Context) {
+	slog.Debug("collector loop running", "sources", len(c.dataSources))
+	for _, d := range c.dataSources {
+		slog.Debug("collecting data from data source", "source", d)
+		if err := c.executeOneDataSource(ctx, d); err != nil {
+			slog.Warn("failed to collect data from data source", "source", d, "error", err)
+		} else {
+			slog.Debug("collected data from data source", "source", d)
+		}
+	}
+	slog.Debug("collector loop complete", "sources", len(c.dataSources))
+}
+
+// executeOneDataSource runs a single iteration of the main loop for a specific data source instance.
+func (c *Collector) executeOneDataSource(ctx context.Context, dataSource DataSource) (err error) {
+	// Increment the generation id
+	dataSource.IncrGenerationID()
+
+	// Get the list of resource pools for this data source
+	pools, err := c.collectResourcePools(ctx, dataSource)
+	if err != nil {
+		return fmt.Errorf("failed to collect resource pools: %w", err)
+	}
+
+	// Get the list of resources for this data source
+	_, err = c.collectResources(ctx, dataSource, pools)
+	if err != nil {
+		return fmt.Errorf("failed to collect resources: %w", err)
+	}
+
+	// TODO: purge stale record
+
+	// TODO: persist data source info
```

Comment:
Another reason not to fail with errors in the for loops. We don´t want the generation id to be greater than the generation id of the data source, in the DB I mean

Reply:
Having a generation id greater than the data source doesn't have to be a problem.  If it is less than the data source than it will be purged.  

I'll think about this a bit and revisit in another PR when I implement the purging.

Reply:
I know, but it is not for us. Support people will have to interact with this. At some point, something could happen. A support person will look at the DB and get very confused (based on training and experience) at seeing rows with generation id greater than the one for the data source.

But yeah, let's discuss this in another pr